File: 2015/Day6.rb
# keywords: 2d array, grid

require 'pp'
data = File.readlines('Day6_input.txt')
# data = File.readlines("Day6_test.txt")

# Ref for 1000 x 1000 array of 0's
# See https://www.thoughtco.com/two-dimensional-arrays-in-ruby-2907737

def switch1(rect, grid, mode)
  (rect[0]..rect[2]).each do |x|
    (rect[1]..rect[3]).each do |y|
      n = mode < 0 ? (grid[y][x] + mode).abs : mode
      grid[y][x] = n
    end
  end
end

def part1(data)
  size = 1000
  lights = Array.new(size) { Array.new(size, 0) }
  c = 0

  data.each do |line|
    # puts line
    rect = []
    s = 0
    a = line.split('through')[0].split(',')[0].split(' ')
    if a.length == 2
      rect[0] = a[1].strip.to_i
      s = -1
    else
      rect[0] = a[2].strip.to_i
      s = if a[1].include?('on')
            1
          else
            0
          end
    end
    rect[1] = line.split('through')[0].split(',')[1].strip.to_i
    rect[2] = line.split('through')[1].split(',')[0].strip.to_i
    rect[3] = line.split('through')[1].split(',')[1].strip.to_i

    switch1(rect, lights, s)
  end

  lights.each do |a|
    c += (a.count 0)
  end
  puts "Part 1 answer: #{1000000 - c}"
end
part1(data)

#
# PART 2
#
def switch2(rect, grid, s)
  (rect[0]..rect[2]).each do |x|
    (rect[1]..rect[3]).each do |y|
      grid[y][x] = grid[y][x] + s
      if grid[y][x] < 0
        grid[y][x] = 0
      end
    end
  end
end

def count2(grid)
  total = 0
  (0..999).each do |x|
    (0..999).each do |y|
      total += grid[y][x]
    end
  end
  total
end

def part2(data)
  size = 1000
  lights = Array.new(size) { Array.new(size, 0) }
  c = 0

  data.each do |line|
    rect = []
    s = 0
    a = line.split('through')[0].split(',')[0].split(' ')
    if a.length == 2
      rect[0] = a[1].strip.to_i
      s = 2
    else
      rect[0] = a[2].strip.to_i
      s = if a[1].include?('on')
            1
          else
            -1
          end
    end
    rect[1] = line.split('through')[0].split(',')[1].strip.to_i
    rect[2] = line.split('through')[1].split(',')[0].strip.to_i
    rect[3] = line.split('through')[1].split(',')[1].strip.to_i

    switch2(rect, lights, s)
  end

  b = 0
  lights.flatten.each { |el| b += el }
  puts "Part 2 answer: #{b}"
end
part2(data)

File: 2015/day6.py
import numpy as np

dim = 1000
lights = np.zeros( (dim, dim) )

# # Part 1
# for line in open("day6_input.txt").readlines():

#   parts = line.split(' ')
#   if len(parts) == 4:
#     action = 't'
#     sX, sY = parts[1].split(',')
#     eX, eY = parts[3].split(',')
#   else:
#     action = parts[1]
#     sX, sY = parts[2].split(',')
#     eX, eY = parts[4].split(',')
  
#   for row in range(int(sY), int(eY)+1):
#     for col in range(int(sX), int(eX)+1):
#       # print("row:", row, "col:", col)
#       if action == 't':
#         lights[row, col] = (lights[row, col] + 1) % 2
#       elif action == 'on':
#         lights[row, col] = 1
#       else:
#         lights[row, col] = 0

# l = 0
# for r in range(dim):
#   for c in range(dim):
#     if lights[r, c] == 1:
#       l += 1

# print("Part 1:",l)

# Part 2
for line in open("day6_input.txt").readlines():
  b = 0
  parts = line.split(' ')
  # print("parts = ", parts)
  if len(parts) == 4:
    b = 2
    sX, sY = parts[1].split(',')
    eX, eY = parts[3].split(',')
  else:
    if parts[1].strip() == 'on':
      b = 1
    else:
      b = -1
    sX, sY = parts[2].split(',')
    eX, eY = parts[4].split(',')
  
  for row in range(int(sY), int(eY)+1):
    for col in range(int(sX), int(eX)+1):
      # print("row:", row, "col:", col, "brightness:", b)
      lights[row, col] += b
      if lights[row, col] < 0:
        lights[row, col] = 0

l = 0
for r in range(dim):
  for c in range(dim):
    l += lights[r, c]

print(int(l))
# print(lights)
